docs/manpages/smbsh.1
.\" This manpage has been automatically generated by docbook2man-spec
.\" from a DocBook document.  docbook2man-spec can be found at:
.\" <http://shell.ipoline.com/~elmert/hacks/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "SMBSH" "1" "23 February 2001" "" ""
.SH NAME
smbsh \- Allows access to Windows NT filesystem  using UNIX commands
.SH SYNOPSIS
.sp
\fBsmbsh\fR
.SH "DESCRIPTION"
.PP
This tool is part of the  Samba <URL:samba.7.html> suite.
.PP
\fBsmbsh\fR allows you to access an NT filesystem 
using UNIX commands such as \fBls\fR, \fB egrep\fR, and \fBrcp\fR. You must use a 
shell that is dynmanically linked in order for \fBsmbsh\fR 
to work correctly.
.PP
To use the \fBsmbsh\fR command, execute \fB smbsh\fR from the prompt and enter the username and password 
that authenticate you to the machine running the Windows NT 
operating system.
.PP
.sp
.nf
	system% \fBsmbsh\fR
	Username: \fBuser\fR
	Password: \fBXXXXXXX\fR
	
.sp
.fi
.PP
Any dynamically linked command you execute from 
this shell will access the \fI/smb\fR directory 
using the smb protocol. For example, the command \fBls /smb
\fRwill show all the machines in your workgroup. The command 
\fBls /smb/<machine-name>\fR will show the share 
names for that machine. You could then, for example, use the \fB cd\fR command to change directories, \fBvi\fR to 
edit files, and \fBrcp\fR to copy files.
.SH "VERSION"
.PP
This man page is correct for version 2.2 of 
the Samba suite.
.SH "BUGS"
.PP
\fBsmbsh\fR works by intercepting the standard 
libc calls with the dynamically loaded versions in \fI smbwrapper.o\fR. Not all calls have been "wrapped", so 
some programs may not function correctly under \fBsmbsh
\fR\&.
.PP
Programs which are not dynamically linked cannot make 
use of \fBsmbsh\fR's functionality. Most versions 
of UNIX have a \fBfile\fR command that will 
describe how a program was linked.
.SH "SEE ALSO"
.PP
\fBsmbd(8)\fR <URL:smbd.8.html>, 
smb.conf(5) <URL:smb.conf.5.html>
.SH "AUTHOR"
.PP
The original Samba software and related utilities 
were created by Andrew Tridgell. Samba is now developed
by the Samba Team as an Open Source project similar 
to the way the Linux kernel is developed.
.PP
The original Samba man pages were written by Karl Auer. 
The man page sources were converted to YODL format (another 
excellent piece of Open Source software, available at
ftp://ftp.icce.rug.nl/pub/unix/ <URL:ftp://ftp.icce.rug.nl/pub/unix/>) and updated for the Samba 2.0 
release by Jeremy Allison. The conversion to DocBook for 
Samba 2.2 was done by Gerald Carter


NEWS
Changes in release 0.3d

 * add krb5-config

 * fix a bug in 3des gss-api mechanism, making it compatible with the
   specification and the MIT implementation

 * make telnetd only allow a specific list of environment variables to
   stop it from setting `sensitive' variables

 * try to use an existing libdes

 * lib/krb5, kdc: use correct usage type for ap-req messages.  This
   should improve compatability with MIT krb5 when using 3DES
   encryption types

 * kdc: fix memory allocation problem

 * update config.guess and config.sub

 * lib/roken: more stuff implemented

 * bug fixes and portability enhancements

Changes in release 0.3c

 * lib/krb5: memory caches now support the resolve operation

 * appl/login: set PATH to some sane default

 * kadmind: handle several realms

 * bug fixes (including memory leaks)

Changes in release 0.3b

 * kdc: prefer default-salted keys on v5 requests

 * kdc: lowercase hostnames in v4 mode

 * hprop: handle more types of MIT salts

 * lib/krb5: fix memory leak

 * bug fixes

Changes in release 0.3a:

 * implement arcfour-hmac-md5 to interoperate with W2K

 * modularise the handling of the master key, and allow for other
   encryption types. This makes it easier to import a database from
   some other source without having to re-encrypt all keys.

 * allow for better control over which encryption types are created

 * make kinit fallback to v4 if given a v4 KDC

 * make klist work better with v4 and v5, and add some more MIT
   compatibility options

 * make the kdc listen on the krb524 (4444) port for compatibility
   with MIT krb5 clients

 * implement more DCE/DFS support, enabled with --enable-dce, see
   lib/kdfs and appl/dceutils

 * make the sequence numbers work correctly

 * bug fixes

Changes in release 0.2t:

 * bug fixes

Changes in release 0.2s:

 * add OpenLDAP support in hdb

 * login will get v4 tickets when it receives forwarded tickets

 * xnlock supports both v5 and v4

 * repair source routing for telnet

 * fix building problems with krb4 (krb_mk_req)

 * bug fixes

Changes in release 0.2r:

 * fix realloc memory corruption bug in kdc

 * `add --key' and `cpw --key' in kadmin

 * klist supports listing v4 tickets

 * update config.guess and config.sub

 * make v4 -> v5 principal name conversion more robust

 * support for anonymous tickets

 * new man-pages

 * telnetd: do not negotiate KERBEROS5 authentication if there's no keytab.

 * use and set expiration and not password expiration when dumping
   to/from ka server databases / krb4 databases

 * make the code happier with 64-bit time_t

 * follow RFC2782 and by default do not look for non-underscore SRV names

Changes in release 0.2q:

 * bug fix in tcp-handling in kdc

 * bug fix in expand_hostname

Changes in release 0.2p:

 * bug fix in `kadmin load/merge'

 * bug fix in krb5_parse_address

Changes in release 0.2o:

 * gss_{import,export}_sec_context added to libgssapi

 * new option --addresses to kdc (for listening on an explicit set of
   addresses)

 * bug fixes in the krb4 and kaserver emulation part of the kdc

 * other bug fixes

Changes in release 0.2n:

 * more robust parsing of dump files in kadmin
 * changed default timestamp format for log messages to extended ISO
   8601 format (Y-M-DTH:M:S)
 * changed md4/md5/sha1 APIes to be de-facto `standard'
 * always make hostname into lower-case before creating principal
 * small bits of more MIT-compatability
 * bug fixes

Changes in release 0.2m:

 * handle glibc's getaddrinfo() that returns several ai_canonname

 * new endian test

 * man pages fixes

Changes in release 0.2l:

 * bug fixes

Changes in release 0.2k:

 * better IPv6 test

 * make struct sockaddr_storage in roken work better on alphas

 * some missing [hn]to[hn]s fixed.

 * allow users to change their own passwords with kadmin (with initial
   tickets)

 * fix stupid bug in parsing KDC specification

 * add `ktutil change' and `ktutil purge'

Changes in release 0.2j:

 * builds on Irix

 * ftpd works in passive mode

 * should build on cygwin

 * work around broken IPv6-code on OpenBSD 2.6, also add configure
   option --disable-ipv6

Changes in release 0.2i:

 * use getaddrinfo in the missing places.

 * fix SRV lookup for admin server

 * use get{addr,name}info everywhere.  and implement it in terms of
   getipnodeby{name,addr} (which uses gethostbyname{,2} and
   gethostbyaddr)

Changes in release 0.2h:

 * fix typo in kx (now compiles)

Changes in release 0.2g:

 * lots of bug fixes:
   * push works
   * repair appl/test programs
   * sockaddr_storage works on solaris (alignment issues)
   * works better with non-roken getaddrinfo
   * rsh works
   * some non standard C constructs removed

Changes in release 0.2f:

 * support SRV records for kpasswd
 * look for both _kerberos and krb5-realm when doing host -> realm mapping

Changes in release 0.2e:

 * changed copyright notices to remove `advertising'-clause.
 * get{addr,name}info added to roken and used in the other code
   (this makes things work much better with hosts with both v4 and v6
    addresses, among other things)
 * do pre-auth for both password and key-based get_in_tkt
 * support for having several databases
 * new command `del_enctype' in kadmin
 * strptime (and new strftime) add to roken
 * more paranoia about finding libdb
 * bug fixes

Changes in release 0.2d:

 * new configuration option [libdefaults]default_etypes_des
 * internal ls in ftpd builds without KRB4
 * kx/rsh/push/pop_debug tries v5 and v4 consistenly
 * build bug fixes
 * other bug fixes

Changes in release 0.2c:

 * bug fixes (see ChangeLog's for details)

Changes in release 0.2b:

 * bug fixes
 * actually bump shared library versions

Changes in release 0.2a:

 * a new program verify_krb5_conf for checking your /etc/krb5.conf
 * add 3DES keys when changing password
 * support null keys in database
 * support multiple local realms
 * implement a keytab backend for AFS KeyFile's

examples/libsmbclient/smbwrapper/smbsh.c
/* 
   Unix SMB/Netbios implementation.
   Version 2.0
   SMB wrapper functions - frontend
   Copyright (C) Andrew Tridgell 1998
   Copyright (C) Derrell Lipman 2003-2005
   
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <limits.h>
#include <string.h>
#include <libsmbclient.h>
#include "bsd-strlfunc.h"

#ifndef FALSE
# define        FALSE   (0)
# define        TRUE    (! FALSE)
#endif

static void smbsh_usage(void)
{
	printf("smbsh [options] [command [args] ...]\n\n");
        printf(" -p prepend library name\n");
        printf(" -a append library name\n");
        printf(" -n");
	printf(" -W workgroup\n");
	printf(" -U username\n");
	printf(" -P prefix\n");
	printf(" -R resolve order\n");
	printf(" -d debug level\n");
	printf(" -l logfile\n");
	printf(" -L libdir\n");
	exit(0);
}

int main(int argc, char *argv[])
{
	char *p, *u;
	char *libd = ".";
	char line[PATH_MAX], pre[PATH_MAX], post[PATH_MAX];
	int opt;
        int no_ask = 0;
        struct stat statbuf;
	extern char *optarg;
	extern int optind;

        *pre = *post = '\0';

	while ((opt = getopt(argc, argv, "p:a:d:nL:W:U:h")) != -1) {
		switch (opt) {
                case 'p':       /* prepend library before smbwrapper.so */
                        if (*pre != '\0')
                                smbw_strlcat(pre, " ", sizeof(pre));
                        smbw_strlcat(pre, optarg, sizeof(pre));
                        break;
                        
                case 'a':       /* append library after smbwrapper.so */
                        smbw_strlcat(post, " ", sizeof(post));
                        smbw_strlcat(post, optarg, sizeof(post));
                        break;
                        
                case 'd':
                        setenv("DEBUG", optarg, TRUE);
                        break;

                case 'n':       /* don't ask for username/password */
                        no_ask++;
                        break;

		case 'L':
			libd = optarg;
			break;

		case 'W':
			setenv("WORKGROUP", optarg, TRUE);
			break;

		case 'U':
			p = strchr(optarg,'%');
			if (p) {
				*p=0;
				setenv("PASSWORD", p+1, TRUE);
			}
			setenv("USER", optarg, TRUE);
			break;

		case 'h':
		default:
			smbsh_usage();
		}
	}


        if (! no_ask) {
                if (!getenv("USER")) {
                        printf("Username: ");
                        u = fgets(line, sizeof(line)-1, stdin);
                        setenv("USER", u, TRUE);
                }
                
                if (!getenv("PASSWORD")) {
                        p = getpass("Password: ");
                        setenv("PASSWORD", p, TRUE);
                }
        }

        smbw_strlcpy(line, pre, PATH_MAX - strlen(line));
        smbw_strlcat(line, " ", sizeof(line));
        smbw_strlcat(line, libd, sizeof(line));
        smbw_strlcat(line, "/smbwrapper.so", sizeof(line));
        smbw_strlcat(line, post, sizeof(line));
	setenv("LD_PRELOAD", line, TRUE);
        setenv("LD_BIND_NOW", "true", TRUE);

	snprintf(line,sizeof(line)-1,"%s/smbwrapper.32.so", libd);

	if (stat(line, &statbuf) == 0 && S_ISREG(statbuf.st_mode)) {
		snprintf(line, sizeof(line)-1,
                         "%s/smbwrapper.32.so:DEFAULT", libd);
		setenv("_RLD_LIST", line, TRUE);
		snprintf(line, sizeof(line)-1,
                         "%s/smbwrapper.so:DEFAULT", libd);
		setenv("_RLDN32_LIST", line, TRUE);
	} else {
		snprintf(line,sizeof(line)-1,"%s/smbwrapper.so:DEFAULT", libd);
		setenv("_RLD_LIST", line, TRUE);
	}

        if (optind < argc) {
                execvp(argv[optind], &argv[optind]);
        } else {
                char *shellpath = getenv("SHELL");

                setenv("PS1", "smbsh$ ", TRUE);

                if(shellpath) {
			execl(shellpath,"smbsh", NULL);
                } else {
                        setenv("SHELL", "/bin/sh", TRUE);
			execl("/bin/sh", "smbsh", NULL);
                }
	}
	printf("launch failed!\n");
	return 1;
}	
